### PR TITLE
use gdal to reproject the asset geometry instead of the asset bbox

### LINKED ISSUE #28: Handling dateline crossing geometry
I'm running into a problem when creating a STAC Item from an image, when the image crosses the dateline.

I've looked at the code, and obviously the original author realised this would be a problem, as there is a TODO comment saying `# TODO: Handle dateline crossing geometry` at https://github.com/developmentseed/rio-stac/blob/master/rio_stac/stac.py#L23

Do you have any suggestions how to handle this? I'd be very happy to do a PR, but this is my first time dealing with things spanning the dateline and I'm not sure how best to handle it.

### PR DESCRIPTION
**Pull request by vincentsarago**
closes #28 

This PR handles dateline crossing assets by using GDAL to handle the reprojection and splitting of the geometry

### CODE REVIEW

File: rio_stac/stac.py [hunk 1]
```diff
@@ -41,19 +42,19 @@ def get_metadata(
...
     metadata: Dict[str, Any] = {}
 
-    # To Do: handle non-geo data
     if src_dst.crs is not None:
-        bbox = warp.transform_bounds(
-            src_dst.crs, "epsg:4326", *src_dst.bounds, densify_pts=21
-        )
+        src_geom = bbox_to_geom(src_dst.bounds)
+        geom = warp.transform_geom(src_dst.crs, "epsg:4326", src_geom)
+        bbox = feature_bounds(geom)
```
Comment (vincentsarago) [inline rio_stac/stac.py:48]: 1. construct geometry using dataset bounds in input projection

2. retroject (and split if needed) the geometry in WGS84

3. get the bbox using rasterio function

**Comment by robintw** [comment]
Unfortunately this doesn't seem to properly solve this problem @vincentsarago 

I've run the code in this PR on an image that crosses the anti-meridian, and I get an item that looks like this:

```
{
    "type": "Feature",
    "stac_version": "1.0.0",
    "id": "4096_0_10.0_01_-3_185",
    "properties": {
        "proj:epsg": 32601,
        "proj:geometry": {
            "type": "Polygon",
            "coordinates": [
                [
                    [
                        377120.0,
                        7618560.0
                    ],
                    [
                        377120.0,
                        7577600.0
                    ],
                    [
                        418080.0,
                        7577600.0
                    ],
                    [
                        418080.0,
                        7618560.0
                    ],
                    [
                        377120.0,
                        7618560.0
                    ]
                ]
            ]
        },
        "proj:bbox": [
            377120.0,
            7577600.0,
            418080.0,
            7618560.0
        ],
        "proj:shape": [
            4096,
            4096
        ],
        "proj:transform": [
            10.0,
            0.0,
            377120.0,
            0.0,
            -10.0,
            7618560.0,
            0.0,
            0.0,
            1.0
        ],
        "datetime": "2000-01-01T00:00:00Z"
    },
    "geometry": {
        "type": "Polygon",
        "coordinates": [
            [
                [
                    -179.0180999600879,
                    68.66685731991039
                ],
                [
                    -178.98557986959838,
                    68.29971948512099
                ],
                [
                    -179.97698264496904,
                    68.28494231541386
                ],
                [
                    179.97430985556775,
                    68.65180108822425
                ],
                [
                    180.0,
                    68.65225994557456
                ],
                [
                    180.0,
                    68.65225994554176
                ],
                [
                    180.0,
                    68.65225994454175
                ],
                [
                    -180.0,
                    68.65225994454175
                ],
                [
                    -180.0,
                    68.65225994554176
                ],
                [
                    -180.0,
                    68.65225994561013
                ],
                [
                    -179.0180999600879,
                    68.66685731991039
                ]
            ]
        ]
    },
    "links": [],
    "bbox": [
        -180.0,
        68.28494231541386,
        180.0,
        68.66685731991039
    ],
    "stac_extensions": [
        "https://stac-extensions.github.io/raster/v1.1.0/schema.json",
        "https://stac-extensions.github.io/projection/v1.0.0/schema.json",
        "https://stac-extensions.github.io/eo/v1.0.0/schema.json"
    ]
}
```

You can see here that the bbox still spans from -180 to +180 because the geometry still has entries of -180 and +180 in it.

Unfortunately I can't share the file I'm using, but I hope it will be possible to debug without it given that the item above has the proj extension so gives the original file co-ordinates.

**Comment by vincentsarago** [comment]
@robintw can you privide the image?

**Comment by vincentsarago** [comment]
@robintw thanks for the feedback, 

Sadly this is an issue at GDAL level 😭 

```python
from rasterio import warp, features

bounds = [377120.0, 7577600.0, 418080.0, 7618560.0]

src_geom = {
    "type": "Polygon",
    "coordinates": [
        [
            [bounds[0], bounds[1]],
            [bounds[2], bounds[1]],
            [bounds[2], bounds[3]],
            [bounds[0], bounds[3]],
            [bounds[0], bounds[1]],
        ]
    ],
}
print("Transform Geometry from 32601 to 4326")
geom = warp.transform_geom("epsg:32601", "epsg:4326", src_geom)
print(geom)
print()
print("BBOX of the new geometry")
print(features.bounds(geom))
print()
>> Transform Geometry from 32601 to 4326
>> {'type': 'Polygon', 'coordinates': [[(-179.0180999600879, 68.66685731991039), (-178.98557986959838, 68.29971948512099), (-179.97698264496904, 68.28494231541386), (179.97430985556775, 68.65180108822425), (180.0, 68.65225994557456), (180.0, 68.65225994554176), (180.0, 68.65225994454175), (-180.0, 68.65225994454175), (-180.0, 68.65225994554176), (-180.0, 68.65225994561013), (-179.0180999600879, 68.66685731991039)]]}

>> BBOX of the new geometry
>> (-180.0, 68.28494231541386, 180.0, 68.66685731991039)

print("Transform bounds from 32601 to 4326")
print(warp.transform_bounds("epsg:32601", "epsg:4326", *bounds))
>> Transform bounds from 32601 to 4326
>> (179.97430985556775, 68.28494231541386, -178.98557986959838, 68.66685731991039)
```

I was hoping that using `rasterio. warp.transform_geom` to handle everything but it seems not 😭. That's said The test I added work for other projection 

```python
bounds = [570585.0, 5612085.0, 803115.0, 5847615.0]

src_geom = {
    "type": "Polygon",
    "coordinates": [
        [
            [bounds[0], bounds[1]],
            [bounds[2], bounds[1]],
            [bounds[2], bounds[3]],
            [bounds[0], bounds[3]],
            [bounds[0], bounds[1]],
        ]
    ],
}
print("Transform Geometry from 32660 to 4326")
geom = warp.transform_geom("epsg:32660", "epsg:4326", src_geom)
print(geom)

>> Transform Geometry from 32660 to 4326
>> {'type': 'MultiPolygon', 'coordinates': [[[(177.9985713516777, 50.65633608225091), (178.04632887400888, 52.773743003086366), (180.0, 52.740407675064105), (180.0, 52.74040767501424), (180.0, 50.62198104614029), (177.9985713516777, 50.65633608225091)]], [[(-178.51382291899003, 52.693427503007605), (-178.71784776800322, 50.58185346765347), (-180.0, 50.62198104608874), (-180.0, 50.62198104609561), (-180.0, 52.74040767501347), (-178.51382291899003, 52.693427503007605)]]]}
```

**Comment by vincentsarago** [comment]
@robintw might be worth to open an issue on gdal 🤷 

```json
// test_32601.geojson
{"type":"Polygon","coordinates":[[[377120,7577600],[418080,7577600],[418080,7618560],[377120,7618560],[377120,7577600]]]}
```
```
ogr2ogr -f GeoJSON -s_srs epsg:32601 -t_srs epsg:4326 out_4326.geojson test_32601.geojson
```
```json
// out_4326.geojson
{ "type": "Polygon", "coordinates": [ [ [ -179.018099960087909, 68.666857319910392 ], [ -178.985579869598382, 68.299719485120988 ], [ -179.976982644969041, 68.284942315413858 ], [ 179.974309855567753, 68.651801088224246 ], [ 180.0, 68.652259945574556 ], [ 180.0, 68.652259945541758 ], [ 180.0, 68.652259944541754 ], [ -180.0, 68.652259944541754 ], [ -180.0, 68.652259945541758 ], [ -180.0, 68.652259945610126 ], [ -179.018099960087909, 68.666857319910392 ] ] ] }
```

**Comment by vincentsarago** [comment]
@robintw I'm going to merge this PR, I'm not quite sure it doesn't work for your file 😭 but to me it seems this PR is doing its best to handle the general issue

File: rio_stac/stac.py (hunk 1 above)
Comment (snowman2) [reply rio_stac/stac.py:47]: I recommend using pyproj for this:
https://pyproj4.github.io/pyproj/stable/api/transformer.html#pyproj.transformer.Transformer.transform_bounds

Or rasterio 1.3+ and GDAL 3.4+.

Related:
- https://github.com/rasterio/rasterio/issues/2222
- https://github.com/rasterio/rasterio/issues/2144
- https://github.com/rasterio/rasterio/issues/2313

Comment (vincentsarago) [reply rio_stac/stac.py:47]: thanks @snowman2.
I'd like to avoid to add a dependency (even if I like pyproj). My understanding is that `transform_geom` should handle antimeridian splitting just fine (confirmed by the test) and the file provided by @robintw was kinda a corner case.

Comment (snowman2) [reply rio_stac/stac.py:47]: That makes sense. One drawback of that approach is you miss out on the benefits of using `transform_bounds`: "Transform boundary densifying the edges to account for nonlinear transformations along these edges and extracting the outermost bounds."

To get around that, you could potentially densify the points in the geometry before passing it to GDAL.
